azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -273,6 +273,109 @@ def _handle_pipeline_event(self, event):
                 logger.warning("incoming pipeline event with no handler.  dropping.")
 
 
+# NOTE: This stage could be a candidate for being refactored into some kind of other
+# pipeline-related structure. What's odd about it as a stage is that it doesn't really respond
+# to operations or events so much as it spawns them on a timer.
+# Perhaps some kind of... Pipeline Daemon?
+class SasTokenRenewalStage(PipelineStage):
+    # Amount of time, in seconds, prior to token expiration, when the renewal process will begin
+    DEFAULT_TOKEN_RENEWAL_MARGIN = 120
+
+    def __init__(self):
+        super(SasTokenRenewalStage, self).__init__()
+        self._token_renewal_timer = None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, op):
+        if (
+            isinstance(op, pipeline_ops_base.InitializePipelineOperation)
+            and self.pipeline_root.pipeline_configuration.sastoken
+        ):
+            self._start_renewal_timer()
+            self.send_op_down(op)
+        else:
+            self.send_op_down(op)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _cancel_token_renewal_timer(self):
+        """Cancel and delete any pending renewal timer"""
+        timer = self._token_renewal_timer
+        self._token_renewal_timer = None
+        if timer:
+            logger.debug("Cancelling SAS Token renewal timer")
+            timer.cancel()
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _start_renewal_timer(self):
+        """Begin a renewal timer.
+        When the timer expires, and the token is renewed, a new timer will be set"""
+        self._cancel_token_renewal_timer()
+        # NOTE: The assumption here is that the SasToken has just been refreshed, so there is
+        # approximately 'TTL' seconds until expiration. In practice this could probably be off
+        # a few seconds given processing time. We could make this more accurate if SasToken
+        # objects had a live TTL value rather than a static one (i.e. "there are n seconds
+        # remaining in the lifespan of this token", rather than "this token was intended to live
+        # for n seconds")
```

Comment:
Given the possibility of clock skew, TTL is so vague that you're really splitting hairs here. :)

Reply:
I agree, that's why I didn't upend the implementation to account for this. Figured it would be a good information to document inline however.

---
In reply to: [427411587](https://github.com/Azure/azure-iot-sdk-python/pull/555#discussion_r427411587) [](ancestors = 427411587)